_posts/2017-07-19-runners-driver.md

```diff
@@ -0,0 +1,49 @@
+---
+title: "Concordion | Runners and Drivers"
+heading: "Concordion Presentations"
+description: "Layering test code with Runners and Drivers"
+identifier: "concordion-presentations"
+author: "Nigel Charman"
+homepage: http://tutansblog.blogspot.co.nz/
+---
+
+![Visual representation of Layering test code with Runners and Drivers](_posts/drivers-runners.png)
+
+## Application Drivers
+An Application Driver is responsible for manipulating an application through one of its interfaces (eg. web app, web service, database, desktop application, host, message queue). 
+
+The Driver will normally require a 3rd party library to interface with the application, such as Selenium WebDriver. (In some cases, it is also possible to interface directly with the service layer of the application.)
+
+A basic driver may expose its functionality in terms of actions on the interface - eg. enter first name, click OK, get error message. 
+
+It is often worthwhile to add functionality to group the actions into workflows that represents the business intent, eg. update payee details, make one-off payment.
+
+A common pattern used for Application Drivers is the [Page Object](https://code.google.com/p/selenium/wiki/PageObjects) pattern. This encapsulates the logic for driving a page in one place so that it is simple to change, and provides an easy interface for users.  Example usage might be:
+
+```groovy
+def loginPage = gotoLoginPage()
+def landingPage = loginPage.loginAs(“user123”, “password”)
+def oneOffPaymentPage = landingPage.selectOneOffPayment()
+oneOffPaymentPage.pay(230.50, “06-9952-5439870-01”)
+```
+
+It is good practice to break the test fixtures into separate Action and Page objects. The Action objects implement user actions, such as "Transfer money between accounts". The Page objects abstract operations on the web page, for instance "Click on Transfer button". The acceptance and automated tests should use the Action objects where possible, but may need to use the Page objects for some operations. The Page objects isolate the tester from changes to the web page (for example if the location or name of the Transfer button are changed). The Action objects isolate the tester from higher level changes, such as "Transfer money" being implemented using a different workflow or metaphor (eg drag and drop).
+Runners
+Once you have implemented Application Drivers, they can be reused for a number of use cases, for example:
+
+* running tests,
+* data setup,
+* scripting 
+
+You can use a wide variety of frameworks for running the tests (basic xUnit style, Spec by Example, data-driven, keyword-driven etc). These frameworks run the tests and report test results for the user.
```

Comment:
Please change first sentence to "You can use a wide variety of frameworks for running the tests, including basic xUnit style, Specification by Example (such as Concordion), data-driven, keyword-driven etc."